tests/cases/compiler/recursiveConditionalTypes2.ts

```diff
@@ -0,0 +1,65 @@
+// @strict: true
+// @lib: esnext
+// @noEmit: true
+
+// repro from https://github.com/microsoft/TypeScript/issues/43877
+
+type UnionToIntersection<U> = (
+  U extends any ? (k: U) => unknown : never
+) extends (k: infer I) => unknown
+  ? I
+  : never;
+
+interface ClassSpec {
+  public?: object;
+  private?: object;
+  publicExtends?: Record<string, ClassSpec>;
+  privateExtends?: Record<string, ClassSpec>;
+}
+
+type MaybeMergePrivateSuperSpec<T> = T extends Record<string, ClassSpec>
+  ? MergePrivateSuperSpec<T>
+  : {};
+
+type MaybeMergePrivateSpecs<T extends ClassSpec, U> = U extends ClassSpec
+  ? MergePrivateSpecs<U, T>
+  : T;
+
+type MergePrivateSpecs<T extends ClassSpec, U extends ClassSpec> = {
+  public: T["public"] & U["public"];
+  private: T["private"] & U["private"];
+};
+
+type MergePrivateSuperSpec<T extends Record<string, ClassSpec>> =
+  UnionToIntersection<
+    {
+      [P in keyof T]: SimplifyPrivateSpec<T[P]>;
+    }[keyof T]
+  >;
+
+export type SimplifyPrivateSpec<T extends ClassSpec> = MaybeMergePrivateSpecs<
+  MaybeMergePrivateSpecs<T, MaybeMergePrivateSuperSpec<T["publicExtends"]>>,
+  MaybeMergePrivateSuperSpec<T["privateExtends"]>
+>;
+
+// repro from https://github.com/microsoft/TypeScript/issues/43877#issuecomment-866146516
```

Comment:
this wasn't fixed by https://github.com/microsoft/TypeScript/pull/46429 , later on it started to loop infinitely (I think?) and that was fixed by https://github.com/microsoft/TypeScript/pull/46599